hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala

```diff
@@ -198,16 +198,25 @@ object LowerMatrixIR {
         .mapRows('row.dropFields(entriesField))
 
     case MatrixColsTable(child) =>
-      let (__let_uid = lower(child).getGlobals) {
-        val sortedCols = irRange(0, irArrayLen('__let_uid(colsField)), 1)
-          .map { 'i ~> makeStruct(
-            '_1 -> '__let_uid(colsField)('i),
-            '_2 -> 'i)
-          }
-          .sort(true, onKey = true)
-          .map { 'elt ~> 'elt('_1)}
-
-          makeStruct('rows -> sortedCols, 'global -> '__let_uid.dropFields(colsField))
+      val colKey = child.typ.colKey
+      let(__cols_and_globals = lower(child).getGlobals) {
+        val sortedCols = if (colKey.isEmpty)
+          '__cols_and_globals (colsField)
+        else
+          irRange(0, irArrayLen('__cols_and_globals (colsField)), 1)
+            .map {
+              'i ~> let(__cols_element = '__cols_and_globals (colsField)('i)) {
+                makeStruct(
+                  // key struct
+                  '_1 -> makeStruct(colKey.map(k => Symbol(k) -> '__cols_element (Symbol(k))): _*),
```

Comment:
Why not just `__cols_element.selectFields(colKey: _*)`? See the `MatrixCollectColsByKey` where I did something similar.

Reply:
I think I wrote this because I was concerned about select fields maintaining field order in the child struct, but it doesn't have those semantics. I'll change it.